First push with CustomEvents Destination Action

wtnelso opened the pull request:
<!-- Hello and thank you for contributing to Segment action-destinations! -->

<!-- Before opening your pull request, make sure you have added and ran unit
     tests and tested your change locally. Refer to our testing
     documentation for more information: https://github.com/segmentio/action-destinations/blob/main/docs/testing.md -->

<!-- If you have questions or issues please open a new issue or create a new discussion
     post in Github. -->

_A summary of your pull request, including the what change you're making and why._

## Testing

_Include any additional information about the testing you have completed to
ensure your changes behave as expected. For a speedy review, please check
any of the tasks you completed below during your testing._

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [ ] [If destination is already live] Tested for backward compatibility of destination. **Note:** New required fields are a breaking change.
- [ ] [Segmenters] Tested in the staging environment
- [ ] [Segmenters] [If applicable for this change] Tested for regression with Hadron.

joe-ayoub-segment commented:
Hi @wtnelso thanks for raising this PR! 
I'll schedule some time to review. 

Could you please remove the yarn version upgrade please? 
The PR contains changes to the package.json and yarn.lock files, which looks like it might have been a mistake.

joe-ayoub-segment commented on packages/destination-actions/src/destinations/attentive/index.ts:
```diff
@@ -0,0 +1,48 @@
+    fields: {
+      apiKey: {
+        label: 'API Key',
+        description: 'API key used to authorize sending data to Attentive.',
```
```suggestion
        description: 'Your Attentive API Key.',
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/attentive/customEvents/index.ts:
```diff
@@ -0,0 +1,92 @@
+    },
+    properties: {
+      label: 'Properties',
+      description: 'Any metadata associated with the event.',
```
```suggestion
      description: 'Metadata to associate with the event.',
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/attentive/customEvents/index.ts:
```diff
@@ -0,0 +1,92 @@
+  fields: {
+    type: {
+      label: 'Type',
+      description: 'The type of event. This name is case sensitive. "Order shipped" and "Order Shipped" would be considered different event types.',
```
hi @wtnelso the value at $.event will always be either 'track', 'identify', 'page', 'screen' or 'group'. Is this information useful to you? If not, then this field probably isn't required.

wtnelso replied:
No, $.event is probably not the correct field then. The "Type" field is typically more descriptive, like "Order Shipped", "Review Completed" or "Loyalty Points Changed". Is there a field similar to that on your end?

joe-ayoub-segment commented on packages/destination-actions/src/destinations/attentive/customEvents/index.ts:
```diff
@@ -0,0 +1,92 @@
+    },
+    externalEventId: {
+      label: 'External Event Id',
+      description: 'A unique identifier representing this specific event. A UUID is recommended.',
```
```suggestion
      description: 'A unique identifier for the event.',
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/attentive/customEvents/index.ts:
```diff
@@ -0,0 +1,92 @@
+            '@path': '$.properties'
+        }
+    },
+    externalEventId: {
+      label: 'External Event Id',
+      description: 'A unique identifier representing this specific event. A UUID is recommended.',
+      type: 'string',
+      required: false,
+    },
```
```suggestion
    externalEventId: {
      label: 'External Event Id',
      description: 'A unique identifier representing this specific event. A UUID is recommended.',
      type: 'string',
      required: false,
      default: {
            '@path': '$.messageId'
      }
    },
```

wtnelso replied:
@joe-ayoub-segment will the messagId value be unique everytime?

joe-ayoub-segment commented on packages/destination-actions/src/destinations/attentive/customEvents/index.ts:
```diff
@@ -0,0 +1,92 @@
+    occurredAt: {
+      label: 'Occurred At',
+      description: 'Timestamp of when the action occurred in ISO 8601 format.',
+      type: 'string', // Keep 'string', as <date-time> is not a valid type in TypeScript.
```
```suggestion
      type: 'string', 
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/attentive/customEvents/index.ts:
```diff
@@ -0,0 +1,92 @@
+    },
+    occurredAt: {
+      label: 'Occurred At',
+      description: 'Timestamp of when the action occurred in ISO 8601 format.',
```
```suggestion
      description: 'Timestamp for the event, ISO 8601 format.',
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/attentive/customEvents/index.ts:
```diff
@@ -0,0 +1,92 @@
+    },
+    phone: {
+      label: 'Phone',
+      description: 'Phone number of the user associated with the action. E.164 format is required. This field is required if either email or an externalIdentifier is not provided.',
```
```suggestion
      description: "The user's phone number in E.164 format. One of 'Phone', 'Email' or 'External Event Id' fields is required.",
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/attentive/customEvents/index.ts:
```diff
@@ -0,0 +1,92 @@
+    },
+    email: {
+      label: 'Email',
+      description: 'Email of the user associated with the action. This field is required if either phone or an externalIdentifier is not provided.',
```
```suggestion
      description: 'The user's email address. One of 'Email', 'Phone' or 'External Event Id' fields is required.',
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/attentive/customEvents/index.ts:
```diff
@@ -0,0 +1,92 @@
+      label: 'Phone',
+      description: 'Phone number of the user associated with the action. E.164 format is required. This field is required if either email or an externalIdentifier is not provided.',
+      type: 'string',
+      required: false,
```
Please add this default mapping for phone: 
```
default: {
        '@if': {
          exists: { '@path': '$.properties.phone' },
          then: { '@path': '$.properties.phone' },
          else: { '@path': '$.context.traits.phone' }
        }
}
```

wtnelso replied:
added to the file and committed.

joe-ayoub-segment commented on packages/destination-actions/src/destinations/attentive/customEvents/index.ts:
```diff
@@ -0,0 +1,92 @@
+        '@path': '$.email'
+    }
+    },
+    clientUserId: {
+      label: 'clientUserId',
+      description: '(optional) Your primary ID for a user. This field is required if either phone, email, or a customIdentifier is not provided.',
+      type: 'string',
+      required: false,
+    },
```
```suggestion
    clientUserId: {
      label: 'Client User ID',
      description: "A primary ID for a user. Required if none of 'Phone', 'Email' and 'External Event Id' are not provided.',
      type: 'string',
      required: false,
    },
```

joe-ayoub-segment commented:
hi @wtnelso I just went through the code and would like to make some changes to make the UI simpler. Do you mind granting me permission to push changes to your repo please? joe-ayoub-segment is my Github handle.

joe-ayoub-segment commented:
hi @wtnelso just following up on this. Are you still interested in having these changes deployed?

wtnelso commented:
> hi @wtnelso just following up on this. Are you still interested in having these changes deployed?

Hi @joe-ayoub-segment apologies for the delay. I didn't get a notification when your comments were made. Just seeing these today.

wtnelso commented:
> joe-ayoub-segment

Just added you to my repo

wtnelso commented:
> Hi @wtnelso thanks for raising this PR! I'll schedule some time to review.
> 
> Could you please remove the yarn version upgrade please? The PR contains changes to the package.json and yarn.lock files, which looks like it might have been a mistake.

Hi @joe-ayoub-segment, apologies but would you be able to clarify where I need to make this change?

joe-ayoub-segment commented on package.json:
```diff
@@ -113,6 +113,7 @@
   "dependencies": {
-    "chance": "^1.1.8"
+    "chance": "^1.1.8",
+    "yarn": "^1.22.22"
```
hi @wtnelso this is where the yarn dependency was upgraded. Can this be removed please?

wtnelso replied:
hi @joe-ayoub-segment updated and committed. but to make sure in understand correctly, is this the desired outcome?
"chance": "^1.1.8",
    "yarn": "^1.22.22"

joe-ayoub-segment replied:
Hi @wtnelso I fixed it when I pushed my changes up. 

There should be no changes to the package.json file.

joe-ayoub-segment commented on yarn.lock:
```diff
@@ -17334,6 +17334,11 @@ yarn-install@^1.0.0:
     chalk "^1.1.3"
     cross-spawn "^4.0.2"
 
+yarn@^1.22.22:
+  version "1.22.22"
+  resolved "https://registry.yarnpkg.com/yarn/-/yarn-1.22.22.tgz#ac34549e6aa8e7ead463a7407e1c7390f61a6610"
+  integrity sha512-prL3kGtyG7o9Z9Sv8IPfBNrWTDmXB4Qbes8A9rEzt6wkJV8mUvoirjU0Mp3GGAU06Y0XQyA3/2/RQFVuK7MTfg==
+
```
Also please remove this @wtnelso .

wtnelso replied:
@joe-ayoub-segment I don't see that in my repo under the yarn.lock file.

joe-ayoub-segment replied:
Also there should be no changes to the yarn.lock file. 
I corrected this when I pushed my changes up.

joe-ayoub-segment replied:
It's OK you can ignore it now anyway. Just pull the latest code and you can ignore these comments :). 

The only thing to do now is fix the unit tests.

wtnelso replied:
Thanks!

joe-ayoub-segment commented:
Hi @wtnelso thanks for granting me access to the repo. I pushed some changes which I thought would be useful. 
1. Merged the identifier fields into a single Object field
2. Added some basic validation
3. Added some types for the request JSON

The tests do break - perhaps you could address these, or point out which part of the code is incorrect. 

Best regards,
Joe

wtnelso commented:
> Hi @wtnelso thanks for granting me access to the repo. I pushed some changes which I thought would be useful.
> 
> 1. Merged the identifier fields into a single Object field
> 2. Added some basic validation
> 3. Added some types for the request JSON
> 
> The tests do break - perhaps you could address these, or point out which part of the code is incorrect.
> 
> Best regards, Joe

hi @joe-ayoub-segment, I added the unit tests. Let me know how they look or what else I can do.

joe-ayoub-segment reviewed (changes requested):
Hi @wtnelso there seems to be a missing file being referenced in the tests. 

import { API_URL } from '../../config'

joe-ayoub-segment commented:
hi @wtnelso I'm still getting some issues with running the tests. 

![image](https://github.com/user-attachments/assets/6aafdb42-6a80-4f18-a85f-3fc4ec50ce62)

wtnelso commented:
Hi @joe-ayoub-segment I believe I made the necessary updates for the tests to now pass.

Is there a way I can test on my side to ensure as well?

wtnelso reviewed (commented):
reviewed changes and made updates

joe-ayoub-segment commented:
hi @wtnelso 
Would you be able to schedule a call with me so we can fix up this Destination and get it deployed please?
Here's my Calendly: https://calendly.com/joe_ayoub/ 

I look forward to catching up. 
Joe

wtnelso commented:
Hi @joe-ayoub-segment, i added a new unit test. other than that new one, not sure there are others to add.

One more question. I used the command `yarn cloud test packages/destination-actions/src/destinations/attentive` and get the below and then following error. Is there anything I need to do different next time?
<img width="1009" alt="image" src="https://github.com/user-attachments/assets/c34dd907-0ebb-40c6-be43-d52605358e42">
Is there something else I need to configure on my end?

joe-ayoub-segment commented:
Hi @wtnelso - which folder did you execute the command from? You should be in the repo root folder.

wtnelso commented:
@joe-ayoub-segment I've tried at the attentive folder and its sub folder. packages/destination-actions/src/destinations/attentive

besides this one command, are there other changes needed for this PR? If we're good on this one, I'll begin the work on the other endpoints we want to add next

joe-ayoub-segment commented:
Hi @wtnelso - the new test doesn't work. 
No probs though, just leave it. I'll fix or remove it. 
Feel free to work on the other Actions.

joe-ayoub-segment commented:
Hi @wtnelso I've prepared this PR for deploy. 
If you are going to write any more code (for the other Actions) please do them in a separate branch.

joe-ayoub-segment commented:
We're in a deploy freeze so next deploy scheduled for 3-Dec

joe-ayoub-segment commented:
hi @wtnelso this PR has been deployed. I'll email you next steps.
